<h1 style="color:lime;font-style:italic; font-family:Georgia">1. Importing the Required Libraries:</h1>

In [2]:
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


import warnings
warnings.filterwarnings('ignore')


import os

<h1 style="color:lime;font-style:italic; font-family:Georgia">2. Reading the Datasets:</h1>

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color:orange;">
WAY I : Individually One by One
</h6>

In [3]:
os.getcwd()  #This is the current directory

'g:\\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\\B2B_Shipping_Cost_Accuracy\\exp_notebook'

In [4]:
courier=pd.read_csv(r'G:\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\B2B_Shipping_Cost_Accuracy\b2b_data\Courier Company - Rates.csv')
invoice = pd.read_csv(r'G:\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\B2B_Shipping_Cost_Accuracy\b2b_data\Invoice.csv')
order = pd.read_csv(r'G:\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\B2B_Shipping_Cost_Accuracy\b2b_data\Order Report.csv')
pincode = pd.read_csv(r'G:\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\B2B_Shipping_Cost_Accuracy\b2b_data\pincodes.csv')
sku = pd.read_csv(r'G:\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\B2B_Shipping_Cost_Accuracy\b2b_data\SKU Master.csv')

In [5]:
courier.head()

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [6]:
invoice.head()

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0  1091117222124  2001806232            1.30             121003   
1  1091117222194  2001806273            1.00             121003   
2  1091117222931  2001806408            2.50             121003   
3  1091117223244  2001806458            1.00             121003   
4  1091117229345  2001807012            0.15             121003   

   Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
0            507101    d  Forward charges                 135.0  
1            486886    d  Forward charges                  90.2  
2            532484    d  Forward charges                 224.6  
3            143001    b  Forward charges                  61.3  
4            515591    d  Forward charges                  45.4

In [7]:
order.head()

ExternOrderNo            SKU  Order Qty  Unnamed: 3  Unnamed: 4
0     2001827036  8904223818706        1.0         NaN         NaN
1     2001827036  8904223819093        1.0         NaN         NaN
2     2001827036  8904223819109        1.0         NaN         NaN
3     2001827036  8904223818430        1.0         NaN         NaN
4     2001827036  8904223819277        1.0         NaN         NaN

In [8]:
pincode.head()

Warehouse Pincode  Customer Pincode Zone  Unnamed: 3  Unnamed: 4
0             121003            507101    d         NaN         NaN
1             121003            486886    d         NaN         NaN
2             121003            532484    d         NaN         NaN
3             121003            143001    b         NaN         NaN
4             121003            515591    d         NaN         NaN

In [9]:
sku.head()

SKU  Weight (g)  Unnamed: 2  Unnamed: 3  Unnamed: 4
0  8904223815682         210         NaN         NaN         NaN
1  8904223815859         165         NaN         NaN         NaN
2  8904223815866         113         NaN         NaN         NaN
3  8904223815873          65         NaN         NaN         NaN
4  8904223816214         120         NaN         NaN         NaN

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color:orange;">
WAY II: Reading Multiple Different Files at Once 
</h6>

In [10]:
# Step 1: Define the folder path where the CSV files are stored
# This is the path to the directory where all our CSV files are located.
folder_path = 'G:\\DataScienceXpert_End_to_End_Soltuion_for_Key_Domains\\B2B_Shipping_Cost_Accuracy\\b2b_data'

# Print all files initially present in the folder (useful for checking what files are there)
print("Initially, the folder contains:", os.listdir(folder_path))
print('-'*130)

# Step 2: List all CSV files in the folder
# We filter out only the files that end with '.csv' to focus on reading CSV files.
csv_files = [file for file in os.listdir(folder_path) if file.endswith('.csv')]
print("List of all csv files got collected and those are :",csv_files)
print('-'*130)

# Step 3: Loop through the CSV files and store them in individual DataFrames
# A dictionary 'dataframes' will store each CSV file's data, with the filename (without '.csv') as the key.
dataframes = {}

# Step 4: Process each CSV file
# This loop reads each file, assigns a name based on the filename, and stores the data in 'dataframes' dictionary.
for file in csv_files:
    file_path = os.path.join(folder_path, file)  # Full file path to read the CSV
    
    try:
        # Extract the filename without the '.csv' extension to use as the DataFrame's name
        df_name = file.split('.')[0]  # It extract only 0th indexed part before '.' ie. name of the file without csv
        
        # Print the DataFrame name to ensure it's working correctly
        print(df_name)  # This will print the name of each DataFrame
        
        # Read the CSV file and store it in the 'dataframes' dictionary
        dataframes[df_name] = pd.read_csv(file_path)
    
    except Exception as e:
        # If there is any error during file reading, print the error and continue
        print(f"Error reading {file}: {e}")

# Now each CSV file is stored as a DataFrame in the 'dataframes' dictionary


Initially, the folder contains: ['Courier Company - Rates.csv', 'Invoice.csv', 'Order Report.csv', 'pincodes.csv', 'SKU Master.csv']
----------------------------------------------------------------------------------------------------------------------------------
List of all csv files got collected and those are : ['Courier Company - Rates.csv', 'Invoice.csv', 'Order Report.csv', 'pincodes.csv', 'SKU Master.csv']
----------------------------------------------------------------------------------------------------------------------------------
Courier Company - Rates
Invoice
Order Report
pincodes
SKU Master


In [11]:
dataframes   #This is the dataframe having key as name of the file without .csv extention and values as the table elements.

{'Courier Company - Rates':    fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
 0         29.5              23.6           33              28.3         40.1   
 
    fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
 0              38.9         45.4              44.8         56.6   
 
    fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
 0              55.5         13.6              23.6         20.5   
 
    rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
 0              28.3         31.9              38.9         41.3   
 
    rto_d_additional  rto_e_fixed  rto_e_additional  
 0              44.8         50.7              55.5  ,
 'Invoice':           AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
 0    1091117222124  2001806232            1.30             121003   
 1    1091117222194  2001806273            1.00             121003   
 2    1091117222931  2001806408            2.50             12100

In [12]:
dataframes['Courier Company - Rates'].head()

fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [13]:
dataframes['Invoice'].head()

AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
0  1091117222124  2001806232            1.30             121003   
1  1091117222194  2001806273            1.00             121003   
2  1091117222931  2001806408            2.50             121003   
3  1091117223244  2001806458            1.00             121003   
4  1091117229345  2001807012            0.15             121003   

   Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
0            507101    d  Forward charges                 135.0  
1            486886    d  Forward charges                  90.2  
2            532484    d  Forward charges                 224.6  
3            143001    b  Forward charges                  61.3  
4            515591    d  Forward charges                  45.4

In [14]:
dataframes['Order Report'].head()

ExternOrderNo            SKU  Order Qty  Unnamed: 3  Unnamed: 4
0     2001827036  8904223818706        1.0         NaN         NaN
1     2001827036  8904223819093        1.0         NaN         NaN
2     2001827036  8904223819109        1.0         NaN         NaN
3     2001827036  8904223818430        1.0         NaN         NaN
4     2001827036  8904223819277        1.0         NaN         NaN

In [15]:
dataframes['pincodes'].head()

Warehouse Pincode  Customer Pincode Zone  Unnamed: 3  Unnamed: 4
0             121003            507101    d         NaN         NaN
1             121003            486886    d         NaN         NaN
2             121003            532484    d         NaN         NaN
3             121003            143001    b         NaN         NaN
4             121003            515591    d         NaN         NaN

In [16]:
dataframes['SKU Master'].head()

SKU  Weight (g)  Unnamed: 2  Unnamed: 3  Unnamed: 4
0  8904223815682         210         NaN         NaN         NaN
1  8904223815859         165         NaN         NaN         NaN
2  8904223815866         113         NaN         NaN         NaN
3  8904223815873          65         NaN         NaN         NaN
4  8904223816214         120         NaN         NaN         NaN

<hr style="border-top: 3px solid yellow;">

<h1 style="color:lime;font-style:italic; font-family:Georgia">3. Basic Data Inspection:</h1>

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color: aqua;">
Checking Number of Records and Features</h6>

In [17]:
# Lets take the dataframes created individually...
print(f"Number of records in courier table {courier.shape[0]}")
print(f"Number of features in courier table {courier.shape[1]}")
print('-'*130)

print(f"Number of records in invoice table {invoice.shape[0]}")
print(f"Number of features in invoice table {invoice.shape[1]}")
print('-'*130)

print(f"Number of records in order table {order.shape[0]}")
print(f"Number of features in order table {order.shape[1]}")
print('-'*130)

print(f"Number of records in pincode table {pincode.shape[0]}")
print(f"Number of features in pincode table {pincode.shape[1]}")
print('-'*130)

print(f"Number of records in sku(stock keeping unit) table {sku.shape[0]}")
print(f"Number of features in sku(stock keeping unit) table {sku.shape[1]}")

Number of records in courier table 1
Number of features in courier table 20
----------------------------------------------------------------------------------------------------------------------------------
Number of records in invoice table 124
Number of features in invoice table 8
----------------------------------------------------------------------------------------------------------------------------------
Number of records in order table 400
Number of features in order table 5
----------------------------------------------------------------------------------------------------------------------------------
Number of records in pincode table 124
Number of features in pincode table 5
----------------------------------------------------------------------------------------------------------------------------------
Number of records in sku(stock keeping unit) table 66
Number of features in sku(stock keeping unit) table 5


<hr style="border-top: 3px solid blue;">

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color: #7DF9FF;">
Checking Columns/Features Present</h6>

In [18]:
print('Features/Columns in Courier Table are:\n',courier.columns)
print('-'*130)

print('Features/Columns in Invoice Table are:\n',invoice.columns)
print('-'*130)

print('Features/Columns in Order Table are:\n',order.columns)
print('-'*130)

print('Features/Columns in Pincode Table are:\n',pincode.columns)
print('-'*130)

print('Features/Columns in SKU Table are:\n',sku.columns)


Features/Columns in Courier Table are:
 Index(['fwd_a_fixed', 'fwd_a_additional', 'fwd_b_fixed', 'fwd_b_additional',
       'fwd_c_fixed', 'fwd_c_additional', 'fwd_d_fixed', 'fwd_d_additional',
       'fwd_e_fixed', 'fwd_e_additional', 'rto_a_fixed', 'rto_a_additional',
       'rto_b_fixed', 'rto_b_additional', 'rto_c_fixed', 'rto_c_additional',
       'rto_d_fixed', 'rto_d_additional', 'rto_e_fixed', 'rto_e_additional'],
      dtype='object')
----------------------------------------------------------------------------------------------------------------------------------
Features/Columns in Invoice Table are:
 Index(['AWB Code', 'Order ID', 'Charged Weight', 'Warehouse Pincode',
       'Customer Pincode', 'Zone', 'Type of Shipment', 'Billing Amount (Rs.)'],
      dtype='object')
----------------------------------------------------------------------------------------------------------------------------------
Features/Columns in Order Table are:
 Index(['ExternOrderNo', 'SKU', 'Order Qt

<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- **`Courier Table`**: Contains forward (`fwd_`) and return (`rto_`) fixed and additional charges categorized by tiers (A to E). Likely represents shipping costs based on different zones or package types.
  
- **`Invoice Table`**: Includes key billing details such as tracking (AWB Code), order identification, shipping weight, pincode information, shipment type, and total billing amount. Useful for cost analysis and shipment tracking.

- **`Order Table`**: Focuses on order details like external order numbers, SKUs, and quantities. The presence of unnamed columns suggests missing or irrelevant data that may need cleaning.

- **`Pincode Table`**: Maps warehouse and customer pincodes to their respective zones, potentially aiding in zone-based cost calculations. Unnamed columns again suggest missing or irrelevant data.

- **`SKU Table`**: Lists SKU (product codes) and their respective weights, which are crucial for calculating shipping costs. Unnamed columns here also indicate potential need for data cleanup.

<hr style="border-top: 3px solid blue;">

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color: aqua">
Random Sample of the dataset.</h6>

In [19]:
print("Random Sample from Courier Table:\n")
courier.sample()   # Since Number of Record is only one.

Random Sample from Courier Table:



fwd_a_fixed  fwd_a_additional  fwd_b_fixed  fwd_b_additional  fwd_c_fixed  \
0         29.5              23.6           33              28.3         40.1   

   fwd_c_additional  fwd_d_fixed  fwd_d_additional  fwd_e_fixed  \
0              38.9         45.4              44.8         56.6   

   fwd_e_additional  rto_a_fixed  rto_a_additional  rto_b_fixed  \
0              55.5         13.6              23.6         20.5   

   rto_b_additional  rto_c_fixed  rto_c_additional  rto_d_fixed  \
0              28.3         31.9              38.9         41.3   

   rto_d_additional  rto_e_fixed  rto_e_additional  
0              44.8         50.7              55.5

In [20]:
print("Random Samples from Invoice Table:\n")
invoice.sample(5)

Random Samples from Invoice Table:



AWB Code    Order ID  Charged Weight  Warehouse Pincode  \
122  1091121844806  2001811475            0.50             121003   
113  1091117806263  2001810549            1.86             121003   
70   1091118548333  2001812854            2.94             121003   
109  1091117613962  2001809270            0.68             121003   
50   1091117227816  2001806801            1.35             121003   

     Customer Pincode Zone Type of Shipment  Billing Amount (Rs.)  
122            173212    b  Forward charges                  33.0  
113            302017    d  Forward charges                 179.8  
70             306302    d  Forward charges                 269.4  
109            324005    d  Forward charges                  90.2  
50             248001    b  Forward charges                  89.6

In [21]:
print('Random Sample from Order Table:\n')
order.sample(5)

Random Sample from Order Table:



ExternOrderNo            SKU  Order Qty  Unnamed: 3  Unnamed: 4
42      2001820978  8904223817501        1.0         NaN         NaN
7       2001827036     SACHETS001        1.0         NaN         NaN
198     2001808832  8904223819338        1.0         NaN         NaN
3       2001827036  8904223818430        1.0         NaN         NaN
203     2001808739  8904223816214        1.0         NaN         NaN

In [22]:
print('Random Sample from Pincode Table:\n')
pincode.sample(5)

Random Sample from Pincode Table:



Warehouse Pincode  Customer Pincode Zone  Unnamed: 3  Unnamed: 4
12             121003            248006    b         NaN         NaN
72             121003            302002    b         NaN         NaN
50             121003            248001    b         NaN         NaN
79             121003            307026    b         NaN         NaN
33             121003            140604    b         NaN         NaN

In [23]:
print('Random Sample from SKU Table:\n')
sku.sample(5)

Random Sample from SKU Table:



SKU  Weight (g)  Unnamed: 2  Unnamed: 3  Unnamed: 4
11  8904223818553         115         NaN         NaN         NaN
40  8904223818898         140         NaN         NaN         NaN
27  8904223818997         490         NaN         NaN         NaN
9   8904223818430         165         NaN         NaN         NaN
58  8904223819260         130         NaN         NaN         NaN

In [24]:
# I have noticed that in order , pincode and sku table Unnamed Columns are there having NaN values only...Lets explore it futher using info().

<hr style="border-top: 3px solid blue;">

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color:aqua">
Examine Datatypes and Information about features</h6>

In [25]:
print("Information about the Courier Table is:")
courier.info()

Information about the Courier Table is:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1 entries, 0 to 0
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fwd_a_fixed       1 non-null      float64
 1   fwd_a_additional  1 non-null      float64
 2   fwd_b_fixed       1 non-null      int64  
 3   fwd_b_additional  1 non-null      float64
 4   fwd_c_fixed       1 non-null      float64
 5   fwd_c_additional  1 non-null      float64
 6   fwd_d_fixed       1 non-null      float64
 7   fwd_d_additional  1 non-null      float64
 8   fwd_e_fixed       1 non-null      float64
 9   fwd_e_additional  1 non-null      float64
 10  rto_a_fixed       1 non-null      float64
 11  rto_a_additional  1 non-null      float64
 12  rto_b_fixed       1 non-null      float64
 13  rto_b_additional  1 non-null      float64
 14  rto_c_fixed       1 non-null      float64
 15  rto_c_additional  1 non-null      float64
 16  rto_d_fi

<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- The table has a single row of data.
- All 20 columns are non-null.
- 19 columns are of type `float64`, representing precise cost data.
- Only one column, `fwd_b_fixed`, is of type `int64`.
- The data represents fixed and additional costs for forward (fwd) and return-to-origin (rto) services across categories A to E.


-------------------

In [26]:
print("Information about the Invoice Table is:")
invoice.info()

Information about the Invoice Table is:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AWB Code              124 non-null    int64  
 1   Order ID              124 non-null    int64  
 2   Charged Weight        124 non-null    float64
 3   Warehouse Pincode     124 non-null    int64  
 4   Customer Pincode      124 non-null    int64  
 5   Zone                  124 non-null    object 
 6   Type of Shipment      124 non-null    object 
 7   Billing Amount (Rs.)  124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- The table contains 124 entries with 8 columns.
- All columns have non-null values, meaning there are no missing data points.
- There are 4 columns with integer data types (`AWB Code`, `Order ID`, `Warehouse Pincode`, `Customer Pincode`).
- Two columns are of type `float64` (`Charged Weight` and `Billing Amount (Rs.)`), representing numeric values.
- The `Zone` and `Type of Shipment` columns are categorical data stored as objects.
- The table likely tracks shipping invoices, including details like AWB codes, shipment zones, and billing amounts.

---------------

In [27]:
print("Information about the Order Table is:")
order.info()

Information about the Order Table is:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ExternOrderNo  400 non-null    int64  
 1   SKU            400 non-null    object 
 2   Order Qty      400 non-null    float64
 3   Unnamed: 3     0 non-null      float64
 4   Unnamed: 4     0 non-null      float64
dtypes: float64(3), int64(1), object(1)
memory usage: 15.8+ KB


<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- The table contains 400 entries.
- Columns `ExternOrderNo`, `SKU`, and `Order Qty` have complete data with no missing values.
- Two columns, `Unnamed: 3` and `Unnamed: 4`, contain entirely null values.
- `ExternOrderNo` is an integer, `SKU` is categorical, and `Order Qty` is a float.
- The unnecessary `Unnamed` columns should be removed for better data quality.

In [28]:
# Lets Remove the Unnamed: 3 and Unnamed: 4 Columns.

order.drop(columns=['Unnamed: 3','Unnamed: 4'],inplace=True)  # inplace =True makesures that , columns will get dropped from the original table mentioned.

In [29]:
order.info()  # No Unnecessary Unnamed columns are there.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ExternOrderNo  400 non-null    int64  
 1   SKU            400 non-null    object 
 2   Order Qty      400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


-------------

In [30]:
print('Information about the Pincode Table:\n')
pincode.info()

Information about the Pincode Table:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Warehouse Pincode  124 non-null    int64  
 1   Customer Pincode   124 non-null    int64  
 2   Zone               124 non-null    object 
 3   Unnamed: 3         0 non-null      float64
 4   Unnamed: 4         0 non-null      float64
dtypes: float64(2), int64(2), object(1)
memory usage: 5.0+ KB


<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- The table contains 124 entries with complete data for `Warehouse Pincode`, `Customer Pincode`, and `Zone`.
- Both `Warehouse Pincode` and `Customer Pincode` columns have integer data types.
- The `Zone` column is categorical and contains non-null values.
- Two columns, `Unnamed: 3` and `Unnamed: 4`, are entirely null and do not provide any useful information.
- The unnecessary `Unnamed` columns should be removed to enhance data clarity.

In [31]:
# Lets Remove the Unnamed: 3 and Unnamed: 4 Columns.

pincode.drop(columns=['Unnamed: 3','Unnamed: 4'],inplace=True) 

In [32]:
pincode.info()   # No Unnecessary Unnamed columns are there.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Warehouse Pincode  124 non-null    int64 
 1   Customer Pincode   124 non-null    int64 
 2   Zone               124 non-null    object
dtypes: int64(2), object(1)
memory usage: 3.0+ KB


------------------

In [33]:
print("Information about the SKU Table:\n")
sku.info()


Information about the SKU Table:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   SKU         66 non-null     object 
 1   Weight (g)  66 non-null     int64  
 2   Unnamed: 2  0 non-null      float64
 3   Unnamed: 3  0 non-null      float64
 4   Unnamed: 4  0 non-null      float64
dtypes: float64(3), int64(1), object(1)
memory usage: 2.7+ KB


<h7 style="color:gold;font-style:italic; font-family:Georgia">Observation:</h7>

- The `SKU Table` contains 66 entries with complete data for the `SKU` and `Weight (g)` columns.
- The `SKU` column is categorical, and `Weight (g)` is an integer representing product weight.
- Three columns (`Unnamed: 2`, `Unnamed: 3`, `Unnamed: 4`) are entirely null and do not add value.
- The unnecessary `Unnamed` columns should be removed to improve data quality.

In [34]:
# Lets Remove the Unnamed: 2,Unnamed: 3 and Unnamed: 4 Columns.

sku.drop(columns=['Unnamed: 2','Unnamed: 3','Unnamed: 4'],inplace=True) 

In [35]:
print("Information about the SKU Table:\n")
sku.info()

Information about the SKU Table:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   SKU         66 non-null     object
 1   Weight (g)  66 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.2+ KB


---------------------------------------

<h6 style="font-family: 'Arial'; font-size: 20px; font-style: italic; color:aqua">
Joining the Tables</h6>

The Order table is the master table, as it contains the most records (400) and serves as the central dataset for transactional data, with other tables providing supplementary details.

In [44]:
invoice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AWB Code              124 non-null    int64  
 1   Order ID              124 non-null    int64  
 2   Charged Weight        124 non-null    float64
 3   Warehouse Pincode     124 non-null    int64  
 4   Customer Pincode      124 non-null    int64  
 5   Zone                  124 non-null    object 
 6   Type of Shipment      124 non-null    object 
 7   Billing Amount (Rs.)  124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


In [45]:
order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ExternOrderNo  400 non-null    int64  
 1   SKU            400 non-null    object 
 2   Order Qty      400 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 9.5+ KB


In [55]:
# Joining the Invoice table to Order table using : ExternOrderNo = Order ID

order_invoice = pd.merge(order, invoice, left_on='ExternOrderNo',right_on ='Order ID', how='outer')
order_invoice.drop(columns=['Order ID'],inplace=True)        
order_invoice.rename(columns={"ExternOrderNo":"Order ID"},inplace=True)

# Or We can optimize the same come as :
# order_invoice.drop(columns=['Order ID'], inplace=True).rename(columns={"ExternOrderNo": "Order ID"}, inplace=True)

In [56]:
order_invoice.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Order ID              400 non-null    int64  
 1   SKU                   400 non-null    object 
 2   Order Qty             400 non-null    float64
 3   AWB Code              400 non-null    int64  
 4   Charged Weight        400 non-null    float64
 5   Warehouse Pincode     400 non-null    int64  
 6   Customer Pincode      400 non-null    int64  
 7   Zone                  400 non-null    object 
 8   Type of Shipment      400 non-null    object 
 9   Billing Amount (Rs.)  400 non-null    float64
dtypes: float64(3), int64(4), object(3)
memory usage: 31.4+ KB


In [42]:
invoice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124 entries, 0 to 123
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   AWB Code              124 non-null    int64  
 1   Order ID              124 non-null    int64  
 2   Charged Weight        124 non-null    float64
 3   Warehouse Pincode     124 non-null    int64  
 4   Customer Pincode      124 non-null    int64  
 5   Zone                  124 non-null    object 
 6   Type of Shipment      124 non-null    object 
 7   Billing Amount (Rs.)  124 non-null    float64
dtypes: float64(2), int64(4), object(2)
memory usage: 7.9+ KB


In [ ]:
invoice_pincode = pd.merge(invoice,pincode,left_on='Warehouse Pincode',right_on='Warehouse Pincode',how='inner')